sensors: BMI160 power management

teburd opened the pull request:
Adds the ti building automation sensor board as a shield, and explores adding device power management to the BMI160. By default the power of a sensor should likely be suspended with an application having control over the power of any given sensor.

The PR adds a simple device pm_toggle shell command which lets a shell user call pm_device_runtime_get or put depending on the state of the device.

Example usage in the sensor shell sample app

``` sh
uart:~$ sensor get bmi160@69 accel_xyz
Failed to read sensor: -5
channel idx=3 accel_xyz x =   0.000000 y =   0.000000 z =   0.000000
uart:~$ device pm_toggle bmi160@69
pm_device_runtime_get(bmi160@69)
uart:~$ sensor get bmi160@69 accel_xyz
channel idx=3 accel_xyz x =  -0.333086 y =   1.090154 z =   9.748596
uart:~$ device pm_toggle bmi160@69
pm_device_runtime_put(bmi160@69)
uart:~$ sensor get bmi160@69 accel_xyz
Failed to read sensor: -5
channel idx=3 accel_xyz x =   0.000000 y =   0.000000 z =   0.000000
```

If this is a useful pattern I'm happy to document it in the sensor docs as the defacto way power management with sensors should be done (I think this is the case, but want feedback).

Commit:
shields: Add TI building sensor shield

The building sensor shield has several useful sensors including a 6 axis
inertial sensor, 3 axis magnetometer, optical light sensor, temperature and
humidity sensor, and connectors for offboard temperature sensors.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
sensor: bmi160: Runtime power management support

Add support for device runtime power management. When suspended sampling
is suspended. When active sampling resumes at the configured rate.

When suspended fetch/get will return an errno and 0'ed out samples.

By default the device will start in a suspended state when
PM_DEVICE_RUNTIME is enabled.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
@ceolin I'm curious on your thoughts if this makes sense or not, or if I might've missed something in terms of how device pm should be used. Thanks!

ceolin commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -712,20 +715,33 @@ static int bmi160_sample_fetch(const struct device *dev,
+	int ret = 0;
+	enum pm_device_state pm_state;
+
+	pm_device_busy_set(dev);
```
That is only used if you build without `PM_DEVICE_RUNTIME_EXCLUSIVE` that is enabled by default when `PM_DEVICE_RUNTIME` is selected. That's said it is reasonable to have it if you don't want the device suspending when the system is idle and you were built without device runtime support.

teburd replied:
This is one of those devices that probably should not be turned off unless explicitly told to do so as it takes so long to come back from suspend (100's of ms).

So right, what I think I want is to not suspend this device if triggers are set or samples are being fetched. I need to add the busy set on trigger set as well, and clear it if the func is NULL.

I need to read more on the system idle part of PM I think to make sure that's what I'd want though.

ceolin replied:
If you are exclusive using device runtime power management you don't need the busy flag, it is enough to hold a device reference (`pm_device_runtime_get`). The busy flag is handy if this device can be built without device runtime power management and you don't want the device be suspended when the soc is idle.

teburd replied:
If that's the case I probably need some conditional inclusion of power management for the device based on whether or not PM_DEVICE_RUNTIME is enabled?

I think that's probably fine, so no PM_DEVICE work is done unless runtime power management is enabled. As in don't do anything unless the application is in charge of PM.

I want to make sure I get this right so I can replicate the pattern and document it correctly

ceolin commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -927,6 +966,15 @@ int bmi160_init(const struct device *dev)
+		LOG_ERR("Failed to enabled runtime power management");
+		return -EIO;
+	}
+
```
It is worth to mention that device is at this moment suspended. So, you have to call `pm_device_runtime_get` before effectively use it.

teburd replied:
It probably makes sense to call ```pm_device_runtime_get``` once in the device init so the behavior with or without PM is the same (the device is active) and the application can turn it off whenever it wants to. I'll do that.

ceolin replied:
not sure ... if you do that here the device will never be suspended since the reference count will be unbalanced (unless that is exactly what you want).

The common operation is to call runtime_get / runtime_put when it is needed, if you can't afford the latency you have to keep it active.

teburd replied:
Yeah on second thought I think the init into a suspended state is the right behavior, it might need to be accounted for in some samples though if I start enabling PM by default in them (like the sensor shell)

ceolin reviewed (dismissed):
changes make sense for me.

teburd commented:
@ceolin Thanks for the feedback, it was really helpful. I need to tweak this a bit I think and review some docs again to consider the implications of the system idle scenario.

I'd like to then maybe, with @MaureenHelm help document the suggested PM strategy for sensors.

Commit:
shell: Provide runtime power management toggling

Adds a device pm_toggle shell command which calls pm_runtime_device
functions put and get to toggle the runtime power state.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

jfischer-no reviewed (changes requested):
Please add minimal documentation for the new shield, what sensors, manufacturer, and so on. Thanks.

teburd commented:
> Please add minimal documentation for the new shield, what sensors, manufacturer, and so on. Thanks.

Added some minimal docs primarily listing the sensors and linking back to the product page. I know some of the shields put a lot of info here but it feels like I'd be duplicate most of the product page by doing so. Happy to write more there if there's specifics you would like to see.

jfischer-no commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -872,6 +887,34 @@ static const struct sensor_driver_api bmi160_api = {
+	}
+	return ret;
+}
+
+
```
Probably one empty line too many.

jfischer-no commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -975,18 +1034,39 @@ int bmi160_init(const struct device *dev)
+	return ret;
+}
+
+
```
Probably one empty line too many.

teburd replied:
Probably! Will clean that up, thanks!

jfischer-no commented on subsys/shell/modules/device_service.c:
```diff
@@ -192,10 +193,48 @@ static int cmd_device_list(const struct shell *sh,
+{
+	const struct device *dev;
+	enum pm_device_state pm_state;
+
```
IMO to complicated ifdefiry, why not just

```c

	if (!IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
		return -ENOTSUP;
	}

	dev = device_get_binding(argv[1]);
	if (dev == NULL) {
		shell_error(sh, "Device unknown (%s)", argv[1]);
		return -ENODEV;
	}
```

teburd replied:
The pm_toggle shell command is completely when the Kconfig is not enabled, happy to leave it around but do the above, there would be a small rom cost add regardless of pm kconfig options then, probably negligible. I’ll wait for others to chime in on this one, I have no strong preference.

jfischer-no replied:
Flash cost is overrated :-)
You could also use `#if defined(CONFIG_PM_DEVICE_RUNTIME)` inside SHELL_STATIC_SUBCMD_SET_CREATE(),
BT shell does this very extensively, https://github.com/zephyrproject-rtos/zephyr/blob/a7c011b0bee4ca3bd72f192641bee56526a07e7a/subsys/bluetooth/shell/bt.c#L3704-L3714

teburd replied:
I think I've sorted this out a bit better, the shell cmd is now a define set or unset in the ifdef with the toggle command function, it feels cleaner anyways

teburd commented:
I don't have a super precise PMU but looking at my power supply amperage clearly shows a nice healthy jump from suspended to active, datasheet says 3uA typical when the sensors are suspended and close to a full 1mA when enabled, I see a 1mA jump when I activate the sensor so seems like.

So seems like it works as expected which is nice

ceolin commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -712,20 +715,31 @@ static int bmi160_sample_fetch(const struct device *dev,
+	int ret = 0;
+	enum pm_device_state pm_state;
+
+	(void)pm_device_state_get(dev, &pm_state);
```
not familiar with this driver but why don't don't you resume (with `pm_device_runtime_get`) do the operation and suspend (or just release the reference with `pm_device_runtime_put`) ? Is it because of being time sensitive ?

ceolin commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -913,9 +955,24 @@ int bmi160_init(const struct device *dev)
+
+	int ret = pm_device_runtime_enable(dev);
+
+	if (ret < 0 && ret != -ENOSYS) {
```
just compare with `< 0`. This function no longer return `-ENOSYS`. The documentation is wrong, I'm sending a fix for that.

ceolin commented:
Minor comments, changes look good to me.
